Add remote cover control via homeassistant and handle unavailable and unknown state.

Links2004 opened the pull request:
# What does this implement/fix?

This change allows to control covers managed my homeassistant the same as local covers.
With this its gets very easy to implement push button control for covers.

There are 2 new conditions for the states unavailable and unknown
- `entity.is_unavailable` 
- `entity.is_unknown`

and 2 new triggers:
- `on_unavailable`
- `on_unknown`

The conditions can be implemented by any component by using the class `EntityBase_State`.
`EntityBase_State` implements 2 ways to set the state
- `void set_entity_state(EntityStateType entity_state);`
- `bool handle_state(const char *input);` (parse the state from a string)

`EntityBase_State` has been implemented for all "homeassistant" components.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/4005

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
api:

cover:
  - platform: homeassistant
    id: my_cool_cover
    entity_id: cover.my_cool_cover
    
  - id: dummy_button
    on_multi_click:
      - timing:
          - ON for at most 0.5s
          - OFF for at least 100ms
        then:
          - if:
              condition:
                - lambda: return id(my_cool_cover).current_operation == CoverOperation::COVER_OPERATION_IDLE;
              then:
                - cover.open:  my_cool_cover
              else:
                - cover.stop:  my_cool_cover
      - timing:
          - ON for at least 1s
        then:
          - cover.open:  my_cool_cover
      - timing:
          - ON for at least 1s
          - OFF for at least 100ms
        then:
          - cover.stop:  my_cool_cover
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

https://github.com/esphome/esphome-docs/pull/4005

Links2004 committed:
Add remote cover control via homeassistant.

Links2004 commented:
docu in the works.
I may do the same for switch.

Links2004 committed:
clang-tidy

nagyrobi commented:
Nice!
Other very common entilites like cilmate and rgb/dimmable lights would be useful too.

Links2004 commented:
cilmate and rgb lights are much more complicated since there is no easy way I know of to discover the features.
and there are many combinations possible.

simple one color dimmable or on/off lights are possible relatively easy with no effects for example.

nagyrobi commented:
Do you handle the `unavailable` state somehow?

Links2004 commented:
not directly, the service calls will not work an the HA side if the device is unavailable.

and based on the docs of HA there is no unavailable state in the API used for getting the status.
https://developers.home-assistant.io/docs/core/entity/cover/#states

nagyrobi commented:
Ok. We still need to configure it as a text sensor and compare the string to `unavailable` to make local automations aware of that.

Links2004 commented:
can you post the code/yaml of getting the `unavailable` state?
you may using a non documented HA API feature?

nagyrobi commented:
Nope, all standard. Here I check textually for `opening` and `closing`:
https://deploy-preview-3678--esphome.netlify.app/cookbook/lvgl#cover-status-and-control

`unknown` and `unavailable` works the same, like
```
        state:
            disabled: !lambda return ((0 == x.compare(std::string{"unavailable"})) or (0 == x.compare(std::string{"unknown"})));
```

Links2004 commented:
ok, then the HA API docs are incomplete :O

looks like esphome currently has no component `unavailable` hadling internaly.
https://github.com/esphome/feature-requests/issues/1568

so there is not generic state I can set for the component to indicate that the cover is `unavailable`.

will take a look if a custom condition is possible, like:
```
cover.is_unavailable: my_cool_cover
```

nagyrobi commented:
Don't forget to trigger a state change in these cases too (even if the state itself will be the same, just this custom condition changes)

Links2004 committed:
add handling of unknown and unavailable state and conditions for it

Links2004 commented:
feel free to test.
and yes `unavailable` and `unknown` are seperate.
but when `unavailable` forces `unknown` to be true too.

nagyrobi commented:
The other entities could benefit from this too (sensor, binary sensor etc).

Links2004 commented:
yay, had the same idea.

the question big question is its a good idea to add the unavailable and unknown bools to all entities of esphome as a optional feature, not all components need it but i can see a lot of cases outside of the homeassistant component.
e.g. components that comunicate external
 - Bluetooth
 - Modbus
 - CAN
....

basically every where there is a posibility that the ESP / esphome itself is running but the sensor or device its comunicating with is not ready or present. 

some feedback frome the esphome main devs is needed I think.

nagyrobi commented:
Sure... But that's a more complex thing.

Links2004 commented:
no really adding the bools to the EntityBase class is easy and deduplicates a lot of code.
I for sure will not implement the `unavailable` and `unknown` in one go for all component.

Its more a laying the ground work so that every component dev can reuse / implement it if it makes sense for them.
the default if not implemented in this case will be `unavailable` and `unknown` = false.
and only if the component specific code changes the state its gets usefull for the component.

landonr commented:
I have a lot of these home assistant components built for homeThing. 
I have the [Number component up for review](https://github.com/esphome/esphome/pull/6455) which includes this [HomeAssistantComponent](https://github.com/esphome/esphome/pull/6455/files#diff-a76257bfb481f16e8c550478c9d2202404d32ae0ba77c6eab87993272ee63251) class that's shared between all the components.

Here's my home assistant cover component https://github.com/landonr/esphome-components/blob/main/components/homeassistant_component/cover/HomeAssistantCover.h

Links2004 commented:
interesting had a short look and did not see any handling of the `unavailable` and  or `unknown` state in home assistant.

nagyrobi commented:
If you implement a native esphome cover (say a time based cover), and unplug the node from network, the entity will become unavailable.

landonr commented:
@Links2004 you're right I haven't done any work with `unavailable`. I agree that it's good to figure out now. I think the same solution should work for all component types (switch, light, cover, etc)

@nagyrobi if the ESPHome device loses wifi, what do you suggest doing with the state? Should these home assistant components subscribe to the network status?

landonr commented on esphome/core/helpers.cpp:
```diff
@@ -409,6 +409,21 @@ ParseOnOffState parse_on_off(const char *str, const char *on, const char *off) {
+  if (strcasecmp(str, "unavailable") == 0)
+    return PARSE_COVER_UNAVAILABLE;
+
+  return PARSE_COVER_UNKNOWN;
```
@nagyrobi what do you think about this for unknown / unavailable 
@Links2004

Links2004 replied:
this code parses the info comming from `subscribe_home_assistant_state` how the info is handled it importend to.
we most likely need the same for all other entity types, may a generic func that only detects the strings `unavailable` and `unknown` that home_assistant may sends, so that we can easy chain it in all cases.

landonr replied:
How would we keep track of the state? The PARSE_COVER_UNAVAILABLE case is specific to covers.
Maybe there should be a shared value for unknown/unavailable that exists outside of state? I'm also fine with this implementation, and having an unavailable state for every type

Links2004 replied:
have rewritten the state handling, now a component only needs to implement the `EntityBase_State` class to get `unavailable` and or `unknown`.

setting the state is as easy as calling `this->set_entity_state` or for string handling `this->handle_state`.

condition names have changed to:
- `entity.is_unavailable`
- `entity.is_unknown`

Links2004 committed:
Rewrite entity state handling

Implement base class EntityBase_State to allow easy and generic implementation of the states in components.

Links2004 committed:
implement EntityBase_State for all homeassistant components

Links2004 committed:
publish_state NAN for homeassistant_sensor when unavailable or unknown.

Links2004 committed:
add support for triggers

Links2004 commented:
added trigger support e.g.:

```yaml
binary_sensor:
  - id: !extend btn_boot
    on_click:
      then:
        - if:
            condition:
              - entity.is_unavailable: my_cool_cover
            then:
              - logger.log: my_cool_cover unavailable
        - if:
            condition:
              - entity.is_unknown: my_cool_cover
            then:
              - logger.log: my_cool_cover unknown

cover:
  - platform: homeassistant
    id: my_cool_cover
    entity_id: cover.my_cool_cover
    on_unavailable:
      then:
         - logger.log: tirgger my_cool_cover unavailable
    on_unknown:
      then:
         - logger.log: tirgger my_cool_cover unknown
```

Links2004 committed:
fix ci

landonr commented:
I haven't tested because I don't have a Home Assistant cover (it's in the mail)
Do you have to worry about ignoring the updated from the API? For example if you change the height to different values quickly, Home Assistant might send back old values which you don't want
1. change the height of the cover on ESPHome
2. height sends from ESPHome to Home Assistant
3. Home Assistant entity updates height state
4. ESPHome receives entity height update
I have some code here for handling this https://github.com/esphome/esphome/pull/6455/files#diff-a76257bfb481f16e8c550478c9d2202404d32ae0ba77c6eab87993272ee63251R19-R24

Links2004 commented:
I had no problems in my testing, the HA API sends a update for every 1% in may tests.

look like your code does "optimistic updates" e.g. your `control` function updates the internal esphome state of the entity.

my code does not do this, and the state is only changed when the HA API sends an update.
Which simply avoids the problem.

I dont see why we need "optimistic updates" and if its there I want a conf parameter to disable it ;)
Changing the internal state to what we think may happen can go wrong on so many levels.

e.g.
- if the service call does fail 
- HA entity does control a device with lockout

the HA API will tell us the correct state as soon that the API call is successful no need for any guessing game.

landonr commented:
> I had no problems in my testing, the HA API sends a update for every 1% in may tests.
> 
> look like your code does "optimistic updates" e.g. your `control` function updates the internal esphome state of the entity.
> 
> my code does not do this, and the state is only changed when the HA API sends an update. Which simply avoids the problem.
> 
> I dont see why we need "optimistic updates" and if its there I want a conf parameter to disable it ;) Changing the internal state to what we think may happen can go wrong on so many levels.
> 
> e.g.
> 
> * if the service call does fail
> * HA entity does control a device with lockout
> * the device has special states that happen on service calls
> 
> the HA API will tell us the correct state as soon that the API call is successful no need for any guessing game.

We want optimistic updates because the entity states can be shown and edited on a menu on the ESP and we don't want the number to jump around. 

For example volume control, light brightness / temp, numbers can be changed immediately, but it might take Home Assistant a full second to return the value. In that time the user can change the state to another value, so we want to update Home Assistant and ignore it

Can you share the config for how you're testing this cover?

Links2004 commented:
I dont see a problem with a shared class for all homeassistant sub components.
will makes thinks easier in the future.

at least in my cases there is no delay to see, the log shows the info all in the same second.
and pressing fast did not result in any problems, at on point I got faster then what my actors allow but thats it.
the state where always good.

what hardware are you running homeassistant on, may the device is to slow?

```
========= Button pressed
[18:08:47][D][binary_sensor:036]: 'Boot Button': Sending state ON
========= Button release
[18:08:47][D][binary_sensor:036]: 'Boot Button': Sending state OFF
========= service call
[18:08:47][D][cover:076]: 'my_cool_cover' - Setting
[18:08:47][D][cover:084]:   Position: 0%
========= state update via API
[18:08:47][D][homeassistant.cover:033]: 'cover.eg_schlafzimmer_rollo': Got state closing
[18:08:47][D][cover:170]: 'my_cool_cover' - Publishing:
[18:08:47][D][cover:173]:   Position: 96%
[18:08:47][D][cover:186]:   Current Operation: CLOSING
========= more state update via API
....
========= state update via API
18:10:06][D][homeassistant.cover:033]: 'cover.eg_schlafzimmer_rollo': Got state closing
[18:10:06][D][cover:170]: 'my_cool_cover' - Publishing:
[18:10:06][D][cover:173]:   Position: 94%
[18:10:06][D][cover:186]:   Current Operation: CLOSING
========= Button pressed
[18:10:07][D][binary_sensor:036]: 'Boot Button': Sending state ON
====== update during press and release happend 
[18:10:07][D][homeassistant.cover:048]: 'cover.eg_schlafzimmer_rollo': Got position 93.00%
[18:10:07][D][cover:170]: 'my_cool_cover' - Publishing:
[18:10:07][D][cover:173]:   Position: 93%
[18:10:07][D][cover:186]:   Current Operation: CLOSING
========= Button release
[18:10:08][D][binary_sensor:036]: 'Boot Button': Sending state OFF
========= service call
[18:10:08][D][cover:076]: 'my_cool_cover' - Setting
[18:10:08][D][cover:080]:   Command: STOP
========= state update via API
[18:10:08][D][homeassistant.cover:033]: 'cover.eg_schlafzimmer_rollo': Got state open
[18:10:08][D][cover:170]: 'my_cool_cover' - Publishing:
[18:10:08][D][cover:173]:   Position: 93%
[18:10:08][D][cover:186]:   Current Operation: IDLE
```

Links2004 committed:
add support for toggle and tilt

Note: tilt untested (no device with tilt)

landonr commented:
@Links2004 that timing looks good, you're right it shouldn't be an issue :D I'll update !6455 with the unknown/unavailable code tonight, can you take a look after?

Links2004 committed:
Merge remote-tracking branch 'upsteam/dev' into homeassistant_cover

Links2004 committed:
Merge branch 'dev' into homeassistant_cover

Links2004 committed:
Merge branch 'dev' into homeassistant_cover

Links2004 commented:
I hate to ping people but any feedback from the esphome Team for thes PR is highly appreciated.

landonr commented:
Sorry for the delay - been afk, i'll try to check this out this week!

jesserockz reviewed (changes requested):
A quick review shows way too many different files changed for this. 

This PR should be split up into the different parts.

I also think the unavailable state handling etc should go in `homeassistant` top level folder only as it is only useful for the platforms of it.

Also see the recent addition that uses `get_home_assistant_state` instead of `get_home_assistant_state` to get things like the supported features attribute for parsing and setting the traits correctly.
